49-hvac-current-monitor/firmware/src/main.cpp

```diff
@@ -0,0 +1,330 @@
+            // Sync the note immediately if the percentage change threshold is
+            // exceeded.
+            JAddBoolToObject(req, "sync", true);
+        }
+    }
+    *lastAvgCurrent = avg;
+
+    if (!notecard.sendRequest(req)) {
+        Serial.println("Failed to send note.");
+        return;
+    }
+
+    Serial.println("Done.");
+}
+
+// This callback is called whenever environment variables are fetched. It's
+// called once per var:val pair.
+static void envVarManagerCb(const char *var, const char *val, void *userCtx)
+{
+    if (strcmp(var, "seconds_between_measurements") == 0) {
+        int seconds = atoi(val);
+
+        if (seconds <= 0) {
+            Serial.println("seconds_between_measurements must be a positive "
+                "integer.");
+        }
+        else {
+            // Flush out the measurements before changing any environment
+            // variables.
+            sendMeasurementsNote(measurements, measurementsIdx,
+                msBetweenMeasurements, pctChangeThreshold,
+                &lastAvgCurrent);
+
+            msBetweenMeasurements = seconds * 1000; // Convert seconds to ms.
+            Serial.print("Seconds between measurements set to ");
+            Serial.print(seconds);
+            Serial.println(".");
+        }
+    }
+    else if (strcmp(var, "percentage_change_threshold") == 0) {
+        float threshold = atof(val);
+
+        if (threshold <= 0) {
+            Serial.println("percentage_change_threshold must be a positive "
+                "number.");
+        }
+        else {
+            // Flush out the measurements before changing any environment
+            // variables.
+            sendMeasurementsNote(measurements, measurementsIdx,
+                msBetweenMeasurements, pctChangeThreshold,
+                &lastAvgCurrent);
+
+            pctChangeThreshold = threshold;
+            Serial.print("Percentage change threshold set to ");
+            Serial.print(pctChangeThreshold);
+            Serial.println(".");
+        }
+    }
+}
+
+float calcVrms()
+{
+    static float samples[SAMPLES_PER_CALC];
+
+    int sample = 0;
+    float sum = 0;
+
+    // Read the samples into an array and sum them up so we can compute the DC
+    // offset via an average.
+    for (size_t i = 0; i < SAMPLES_PER_CALC; ++i) {
+        samples[i] = analogRead(ANALOG_PIN);
+        sum += samples[i];
+    }
+
+    // Average the samples to get the DC offset.
+    float offset = sum / SAMPLES_PER_CALC;
+    float sampleWithOffsetRemoved = 0;
+    sum = 0;
+
+    for (size_t i = 0; i < SAMPLES_PER_CALC; ++i) {
+        // Subtract the offset from each sample.
+        sampleWithOffsetRemoved = samples[i] - offset;
+        // Sum the squares of the samples with offset removed.
+        sum += sampleWithOffsetRemoved * sampleWithOffsetRemoved;
+    }
+
+    // Use the sum of squares to compute the RMS voltage.
+    float rms = (sqrt(sum / SAMPLES_PER_CALC) / ADC_RANGE) * ADC_REF_VOLTAGE;
```

Comment:
I get the RMS, but puzzled about the division by `ADC_RANGE` because the RMS can only be positive, up to half the range?

Reply:
So I know that this calculation is somehow right because it agrees with what I see for RMS voltage on my multimeter. I _think_ what's happening is that `ADC_RANGE` being off by a factor of 1/2, as you point out, is canceled out by the fact that `ADC_REF_VOLTAGE` is also off by a factor of 2. But I'll need to think about this one some more...